File: driver/director.py
"""Pure, bounded offline rehearsal and timing proposals for authored moves.

Director samples the canonical :class:`moves.Move`; it has no camera, thread,
storage or network dependency. Findings are sampled engineering checks, not a
physical certification of a gimbal, lens, mounting, scene or resulting shot.
"""

from __future__ import annotations

import math

from . import moves, preflight, response

MAX_VISUAL_SAMPLES = 600
VISUAL_HZ = 20.0
MAX_WAYPOINTS = 200
MAX_TARGET_DURATION = 24.0 * 60.0 * 60.0
SUGGESTION_ATTEMPTS = 8
SPEED_MARGIN = 1.02
SPAN_TOLERANCE_S = 1e-9


def _number(value, name: str) -> float:
    if isinstance(value, bool):
        raise ValueError(f"{name} must be a finite number")
    try:
        got = float(value)
    except (TypeError, ValueError, OverflowError):
        raise ValueError(f"{name} must be a finite number") from None
    if not math.isfinite(got):
        raise ValueError(f"{name} must be a finite number")
    return got


def _span(move: moves.Move) -> float:
    return move.cycle_duration


def _visual_samples(move: moves.Move) -> list[dict]:
    span = _span(move)
    if len(move.waypoints) < 2 or span <= 0:
        return []
    count = min(MAX_VISUAL_SAMPLES,
                max(2, int(math.ceil(span * VISUAL_HZ)) + 1))
    out = []
    for i in range(count):
        t = span * i / (count - 1)
        point = move.sample(t)
        if point is None:
            continue
        zoom = move.sample_zoom(t)
        out.append({
            "time": round(t, 4),
            "pitch": round(point[0], 6),
            "yaw": round(point[1], 6),
            "zoom": None if zoom is None else round(zoom, 6),
        })
    return out


def _beat_record(index: int, waypoint: moves.Waypoint, direction: str,
                 arrival: float, dwell_start: float,
                 dwell_end: float) -> dict:
    return {
        "index": index,
        "name": waypoint.name,
        "direction": direction,
        "arrival": round(arrival, 4),
        "dwell_start": round(dwell_start, 4),
        "dwell_end": round(dwell_end, 4),
        "cue": waypoint.cue,
    }


def _beats_and_events(move: moves.Move) -> tuple[list[dict], list[dict]]:
    arrivals = move.arrival_times()
    beats: list[dict] = []
    events: list[dict] = []
    cue_events: set[tuple[float, int]] = set()

    def add(index: int, direction: str, arrival: float,
            dwell_start: float, dwell_end: float) -> None:
        waypoint = move.waypoints[index]
        beats.append(_beat_record(index, waypoint, direction, arrival,
                                  dwell_start, dwell_end))
        base = {"beat_index": index, "name": waypoint.name,
                "direction": direction}
        events.append({**base, "time": round(arrival, 4), "type": "arrival"})
        if waypoint.dwell > 0:
            events.append({**base, "time": round(dwell_start, 4),
                           "type": "dwell_start"})
            events.append({**base, "time": round(dwell_end, 4),
                           "type": "dwell_end"})
        if waypoint.cue:
            cue_key = (round(arrival, 4), index)
            if cue_key not in cue_events:
                events.append({**base, "time": cue_key[0], "type": "cue"})
                cue_events.add(cue_key)

    for i, arrival in enumerate(arrivals):
        add(i, "forward", arrival, arrival,
            arrival + move.waypoints[i].dwell)

    if move.ping_pong:
        cycle = move.cycle_duration
        for i in range(len(move.waypoints) - 1, -1, -1):
            waypoint = move.waypoints[i]
            dwell_start = cycle - (arrivals[i] + waypoint.dwell)
            dwell_end = cycle - arrivals[i]
            add(i, "reverse", dwell_start, dwell_start, dwell_end)

    event_order = {"arrival": 0, "cue": 1, "dwell_start": 2, "dwell_end": 3}
    events.sort(key=lambda event: (event["time"], event_order[event["type"]],
                                   event["beat_index"]))
    beats.sort(key=lambda beat: (beat["arrival"], beat["index"]))
    return beats, events


def _retimed_for_span(move: moves.Move, duration: float) -> moves.Move:
    divisor = 2.0 if move.ping_pong else 1.0
    return move.retimed(total=duration / divisor)


def _report(move: moves.Move, max_dps: float,
            min_dps: float) -> preflight.Report:
    return preflight.check(move, max_dps=max_dps, min_dps=min_dps)


def _problem_reason(report: preflight.Report) -> str:
    kinds = {finding.kind for finding in report.findings}
    if "assessment" in kinds:
        return "move is outside the bounded offline assessment envelope"
    if "travel" in kinds:
        return "travel violations cannot be repaired by changing timing"
    if "too fast" in kinds and "dead band" in kinds:
        return "no checked duration satisfied both maximum speed and dead-band limits"
    if "too fast" in kinds:
        return "requested timing is too fast"
    if "dead band" in kinds:
        return "requested timing leaves motion below the dead band"
    return "no checked timing proposal is available"


def _suggest(candidate: moves.Move, report: preflight.Report,
             max_dps: float, min_dps: float) -> tuple[moves.Move, preflight.Report] | None:
    current = candidate
    current_report = report
    for _ in range(SUGGESTION_ATTEMPTS):
        fast = [abs(f.worst) / max_dps for f in current_report.findings
                if f.kind == "too fast"]
        slow = [abs(f.worst) / min_dps for f in current_report.findings
                if f.kind == "dead band"]
        lower = max(fast, default=0.0) * SPEED_MARGIN
        upper = min(slow, default=math.inf) / SPEED_MARGIN
        if fast and slow and lower > upper:
            return None
        if fast:
            factor = max(lower, 1.01)
        elif slow:
            factor = min(upper, 0.99)
        else:
            return None
        if not math.isfinite(factor) or factor <= 0:
            return None
        try:
            next_move = current.retimed(factor=factor)
        except ValueError:
            return None
        if (_span(next_move) > MAX_TARGET_DURATION
                or abs(_span(next_move) - _span(current)) < 1e-9):
            return None
        next_report = _report(next_move, max_dps, min_dps)
        if next_report.ok:
            return next_move, next_report
        if any(f.kind in ("travel", "assessment") for f in next_report.findings):
            return None
        current, current_report = next_move, next_report
    return None


def _proposal(move: moves.Move, original: preflight.Report,
              max_dps: float, min_dps: float,
              target_duration: float | None) -> dict:
    requested = target_duration
    candidate = move
    candidate_report = original
    if requested is not None:
        candidate = _retimed_for_span(move, requested)
        candidate_report = _report(candidate, max_dps, min_dps)
        if candidate_report.ok:
            assessed = _span(candidate)
            if not math.isclose(assessed, requested, rel_tol=1e-12,
                                abs_tol=SPAN_TOLERANCE_S):
                return {
                    "status": "suggested", "requested_duration": requested,
                    "assessed_duration": assessed,
                    "move": candidate.to_dict(),
                    "preflight": candidate_report.to_dict(),
                    "reason": (
                        "requested duration is shorter than the move's minimum "
                        "leg timing permits; returning the checked minimum-"
                        "timing alternative"
                    ),
                }
            return {
                "status": "requested_feasible", "requested_duration": requested,
                "assessed_duration": assessed, "move": candidate.to_dict(),
                "preflight": candidate_report.to_dict(), "reason": None,
            }
    elif candidate_report.ok:
        return {
            "status": "none", "requested_duration": None,
            "assessed_duration": _span(move), "move": None,
            "preflight": None, "reason": None,
        }

    if any(f.kind in ("travel", "assessment") for f in candidate_report.findings):
        return {
            "status": "impossible", "requested_duration": requested,
            "assessed_duration": _span(candidate), "move": None,
            "preflight": candidate_report.to_dict(),
            "reason": _problem_reason(candidate_report),
        }

    suggested = _suggest(candidate, candidate_report, max_dps, min_dps)
    if suggested is None:
        return {
            "status": "impossible", "requested_duration": requested,
            "assessed_duration": _span(candidate), "move": None,
            "preflight": candidate_report.to_dict(),
            "reason": _problem_reason(candidate_report),
        }
    suggested_move, suggested_report = suggested
    return {
        "status": "suggested", "requested_duration": requested,
        "assessed_duration": _span(suggested_move),
        "move": suggested_move.to_dict(),
        "preflight": suggested_report.to_dict(),
        "reason": _problem_reason(candidate_report),
    }


def preview(move: moves.Move, max_dps: float = response.MAX_DPS,
            min_dps: float = response.MIN_DPS,
            target_duration: float | None = None) -> dict:
    """Return a bounded canonical preview and a checked, unapplied proposal.

    ``target_duration`` means finite playback duration, or one complete cycle
    for a loop. A ping-pong target therefore retimes its forward half to half
    that value. Every returned candidate is produced solely by ``Move.retimed``.
    """
    if not isinstance(move, moves.Move):
        raise ValueError("move must be a Move")
    move.validate_axis_curves()
    if len(move.waypoints) > MAX_WAYPOINTS:
        raise ValueError(f"Director supports at most {MAX_WAYPOINTS} waypoints")
    try:
        authored_span = _number(_span(move), "move cycle duration")
    except ValueError:
        raise ValueError(
            "move cycle duration must be finite and within 0.."
            f"{MAX_TARGET_DURATION:g} seconds for Director rehearsal"
        ) from None
    if authored_span < 0 or authored_span > MAX_TARGET_DURATION:
        raise ValueError(
            "move cycle duration must be finite and within 0.."
            f"{MAX_TARGET_DURATION:g} seconds for Director rehearsal")
    max_dps = _number(max_dps, "max_dps")
    min_dps = _number(min_dps, "min_dps")
    if min_dps <= 0 or max_dps < min_dps:
        raise ValueError("speed interval must satisfy 0 < min_dps <= max_dps")
    if target_duration is not None:
        target_duration = _number(target_duration, "target_duration")
        if target_duration <= 0 or target_duration > MAX_TARGET_DURATION:
            raise ValueError(
                f"target_duration must be within 0..{MAX_TARGET_DURATION:g} seconds")

    report = _report(move, max_dps, min_dps)
    beats, events = _beats_and_events(move)
    return {
        "samples": _visual_samples(move),
        "beats": beats,
        "events": events,
        "timing": {
            "forward_duration": move.total_duration,
            "programmed_duration": move.playback_duration,
            "loop_uncertain": move.loop,
            "cue_uncertain": move.has_cues,
            "requested_duration": target_duration,
        },
        "preflight": report.to_dict(),
        "proposal": _proposal(move, report, max_dps, min_dps,
                              target_duration),
        "approximation": {
            "visual_sample_limit": MAX_VISUAL_SAMPLES,
            "preflight_sample_limit": preflight.MAX_SAMPLES,
            "leg_intervals": preflight.LEG_INTERVALS,
            "rehearsal_duration_limit": MAX_TARGET_DURATION,
            "target_duration_basis": "finite playback or one loop cycle",
            "scope": ("sampled arithmetic rehearsal only; not physical "
                      "certification or hardware acceptance"),
        },
    }

File: web/director.js
/* Data-only rehearsal. The host's Move.sample owns every pose; this module
   interpolates only the display cursor between returned visual samples. */
(() => {
  'use strict';
  window.createDirector = (root, api) => {
    root.innerHTML=`
      <div class="d-heading"><div><div class="d-kicker">Director / motion rehearsal</div>
        <h2>Give every move a moment.</h2><p>Read the rhythm, rehearse the beats, then choose the timing. Your camera stays still here.</p></div>
        <span class="d-badge">PREVIEW ONLY · NO CAMERA MOVEMENT</span></div>
      <div class="d-layout"><section class="d-stage" aria-label="Offline shot preview">
        <div class="d-stagebar"><strong id="dName">No shot loaded</strong><span id="dViewLabel">Authored path</span></div>
        <div class="d-plot"><canvas id="dPlot" role="img" aria-label="Gimbal angle path. Text positions and timing are available in the beat buttons below."></canvas>
          <div class="d-legend"><span>GIMBAL PATH · PAN × TILT</span><span id="dPose">No positions</span></div></div>
        <div class="d-transport"><div class="d-time"><output id="dTime">00.00 s</output><span id="dEnd">—</span></div>
          <input id="dScrub" type="range" min="0" max="1" step="0.01" value="0" aria-label="Offline rehearsal time" disabled>
          <div class="d-actions"><button id="dPlay" class="primary" disabled>▶ Preview shot</button><button id="dReset" disabled>↤ Start</button><button id="dNext" disabled>Next beat →</button></div>
          <div id="dCue" class="d-cue" hidden><span id="dCueText"></span><div class="d-actions"><button id="dContinue">Continue preview →</button></div></div>
          <p id="dState" class="d-foot" role="status">Add at least two positions in Compose to rehearse a shot.</p></div>
      </section><aside class="d-pace" aria-labelledby="dPaceTitle"><div class="d-kicker">Shot pace</div><h3 id="dPaceTitle">Timing, with intent.</h3>
        <div id="dDuration" class="d-duration">—</div><p id="dTimingNote">Programmed motion time, excluding human cue waits.</p>
        <label for="dTarget">Target motion duration</label><div class="d-input"><input id="dTarget" type="number" min="0.1" max="86400" step="0.1" placeholder="e.g. 12"><span>s</span></div>
        <div class="d-actions"><button id="dFit" disabled>Check timing</button><button id="dAuto" disabled>Find workable timing</button></div>
        <div id="dProposal" class="d-proposal" hidden><strong id="dProposalTitle"></strong><p id="dProposalText"></p><ul id="dChanges"></ul>
          <div class="d-actions"><button id="dPreviewProposal" disabled>Preview proposal</button><button id="dApply" class="primary" disabled>Apply timing</button></div></div>
        <div class="d-actions"><button id="dUndo" hidden>↶ Restore previous timing</button></div>
        <p class="d-foot">Timing scales travel and fixed holds together. Cues still wait for you. No automatic arming, motion, or recording.</p>
      </aside></div>
      <section class="d-score"><div class="d-scorehead"><div><div class="d-kicker">Shot score</div><h3>Your framing beats</h3></div><div class="d-actions"><button id="dEdit" disabled>Edit selected beat ↗</button><button id="dRefresh">Refresh analysis</button></div></div>
        <div id="dBeats" class="d-beats" aria-label="Select a framing beat"></div>
        <div class="d-evidence"><h3 id="dCheckTitle">Path check</h3><ul id="dFindings" class="d-findings"></ul><p id="dLimits" class="d-foot"></p></div>
        <p class="d-foot">This is sampled gimbal-angle arithmetic, not a view through the lens or a physical guarantee. Zoom tracks execute when camera readback permits; this preview does not prove lens response. Verify actual footage before relying on a shot.</p>
      </section>`;
    const el=id=>root.querySelector('#'+id);
    let data=null, shown=null, candidate=null, original=null, undo=null, requestId=0, displayPoints=[];
    let t=0, playing=false, frame=0, last=0, selected=0, waiting=null, consumed=new Set(), applying=false;
    const reducedMotion=window.matchMedia('(prefers-reduced-motion: reduce)');
    let lastDraw=0;
    const end=()=>shown?.samples?.at(-1)?.time||0;
    const seconds=n=>`${Number(n).toFixed(2)} s`;
    function pause(){playing=false;cancelAnimationFrame(frame);el('dPlay').textContent='▶ Preview shot';}
    function clearAnalysis(message){pause();data=shown=candidate=original=null;displayPoints=[];waiting=null;t=0;
      el('dCue').hidden=true;el('dProposal').hidden=true;
      for(const id of ['dPlay','dReset','dNext','dScrub','dFit','dAuto','dEdit','dApply','dPreviewProposal'])el(id).disabled=true;
      el('dBeats').replaceChildren();el('dFindings').replaceChildren();
      el('dCheckTitle').textContent='Path not assessed';el('dLimits').textContent='';
      el('dDuration').textContent=el('dEnd').textContent='—';
      el('dState').textContent=message;display();
    }
    function invalidate(){requestId++;undo=null;el('dUndo').hidden=true;
      clearAnalysis('Draft changed. Refresh analysis to rehearse the current shot.');
      window.copilot?.invalidate();
      window.spatialStudio?.invalidate();
    }
    function pose(){
      const points=displayPoints;if(!points.length)return null;
      let hi=points.findIndex(p=>p.time>=t);if(hi<0)hi=points.length-1;
      const b=points[hi],a=points[Math.max(0,hi-1)],mix=b.time===a.time?0:(t-a.time)/(b.time-a.time);
      return {pitch:a.pitch+(b.pitch-a.pitch)*mix,yaw:a.yaw+(b.yaw-a.yaw)*mix};
    }
    function draw(){
      const canvas=el('dPlot'),rect=canvas.getBoundingClientRect();if(rect.width<1)return;
      const dpr=Math.min(2,window.devicePixelRatio||1),w=rect.width,h=rect.height;
      canvas.width=w*dpr;canvas.height=h*dpr;const c=canvas.getContext('2d');c.scale(dpr,dpr);
      c.clearRect(0,0,w,h);const pts=displayPoints;
      c.strokeStyle='#24333e';c.lineWidth=1;
      for(let x=30;x<w;x+=45){c.beginPath();c.moveTo(x,16);c.lineTo(x,h-26);c.stroke();}
      for(let y=22;y<h-20;y+=40){c.beginPath();c.moveTo(22,y);c.lineTo(w-22,y);c.stroke();}
      if(!pts.length){c.fillStyle='#9aaebc';c.font='13px sans-serif';c.textAlign='center';c.fillText('Your path will appear here',w/2,h/2);return;}
      const ys=pts.map(p=>p.yaw),ps=pts.map(p=>p.pitch);
      const ym=(Math.max(...ys)+Math.min(...ys))/2,pm=(Math.max(...ps)+Math.min(...ps))/2;
      const span=Math.max(Math.max(...ys)-Math.min(...ys),Math.max(...ps)-Math.min(...ps),8)*1.4;
      const scale=Math.min(w-80,h-70)/span;
      const xy=p=>[w/2+(p.yaw-ym)*scale,h/2-(p.pitch-pm)*scale];
      c.strokeStyle='#6ca1ae';c.lineWidth=2;c.beginPath();pts.forEach((p,i)=>{const q=xy(p);i?c.lineTo(...q):c.moveTo(...q);});c.stroke();
      const beats=shown.beats.filter(b=>b.direction==='forward');
      beats.forEach(b=>{const p=pts.reduce((best,p)=>Math.abs(p.time-b.arrival)<Math.abs(best.time-b.arrival)?p:best,pts[0]),q=xy(p);
        c.fillStyle='#172832';c.strokeStyle='#a7cbd1';c.beginPath();c.arc(...q,11,0,Math.PI*2);c.fill();c.stroke();
        c.fillStyle='#deeaee';c.font='10px ui-monospace,monospace';c.textAlign='center';c.fillText(String(b.index+1),q[0],q[1]+3);});
      const p=pose();if(p){const [x,y]=xy(p);c.strokeStyle='#9be7e9';c.fillStyle='#a3e2e4';c.lineWidth=1.5;
        c.beginPath();c.arc(x,y,5,0,Math.PI*2);c.fill();c.beginPath();c.arc(x,y,18,0,Math.PI*2);c.stroke();
        c.strokeStyle='#728c9a';c.beginPath();c.moveTo(24,h-25);c.lineTo(84,h-25);c.stroke();
        c.fillStyle='#9bafbd';c.font='11px ui-monospace,monospace';c.textAlign='left';c.fillText(`${(60/scale).toFixed(1)}°`,24,h-8);}
    }
    function display(force=true){
      el('dScrub').value=t;el('dTime').textContent=seconds(t);
      const p=pose();el('dPose').textContent=p?`PAN ${p.yaw.toFixed(1)}° · TILT ${p.pitch.toFixed(1)}°`:'No positions';
      const beat=(shown?.beats||[]).filter(b=>b.arrival<=t+.0001).at(-1);if(beat)selected=beat.index;
      el('dBeats').querySelectorAll('button').forEach(b=>b.setAttribute('aria-pressed',String(+b.dataset.index===selected)));
      // Keep user-requested playback and its clock accurate, but use a quiet
      // stepped cursor instead of continuous canvas motion when requested.
      const now=performance.now();
      if(force||!reducedMotion.matches||now-lastDraw>=500){draw();lastDraw=now;}
    }
    function seek(value){pause();waiting=null;el('dCue').hidden=true;t=Math.max(0,Math.min(end(),value));
      consumed=new Set((shown?.events||[]).map((e,i)=>e.type==='cue'&&e.time<t?i:-1));display();}
    function tick(now){
      if(!playing)return;const next=Math.min(end(),t+Math.max(0,(now-last)/1000));last=now;
      const cueIndex=(shown.events||[]).findIndex((e,i)=>e.type==='cue'&&!consumed.has(i)&&e.time>=t-.0001&&e.time<=next);
      if(cueIndex>=0){const cue=shown.events[cueIndex];t=cue.time;waiting=cueIndex;pause();
        el('dCueText').textContent=`Cue at P${cue.beat_index+1} · ${cue.name||'Untitled beat'}. Live duration depends on your GO.`;
        el('dCue').hidden=false;display();return;}
      t=next;display(false);if(t>=end()){pause();display();el('dState').textContent='Preview complete. Camera unchanged.';return;}frame=requestAnimationFrame(tick);
    }
    function play(){if(!shown?.samples.length)return;if(playing){pause();return;}if(waiting!==null)return;window.spatialStudio?.pause();
      if(t>=end())seek(0);playing=true;last=performance.now();el('dPlay').textContent='Ⅱ Pause preview';frame=requestAnimationFrame(tick);}
    function renderShown(){
      // Equivalent angle turns keep the chart continuous across +/-180.
      // This is display normalization only, never a new commanded path.
      displayPoints=[];
      for(const point of shown.samples){const previous=displayPoints.at(-1),p={...point};
        if(previous)for(const axis of ['pitch','yaw'])p[axis]=previous[axis]+((p[axis]-previous[axis]+180)%360+360)%360-180;
        displayPoints.push(p);}
      el('dDuration').textContent=seconds(shown.timing.programmed_duration??end());
      el('dTimingNote').textContent=(shown.timing.loop_uncertain?'One loop cycle; live duration is open-ended. ':shown.timing.forward_duration!==shown.timing.programmed_duration?'Round trip, including reverse. ':'Programmed travel and fixed holds. ')+(shown.timing.cue_uncertain?'Plus live cue waits—total shoot time is unknown.':'No cue waits.');
      el('dScrub').max=end();el('dEnd').textContent=seconds(end());el('dBeats').replaceChildren();
      for(const b of shown.beats){if(b.direction!=='forward')continue;const button=document.createElement('button');button.className='d-beat';button.dataset.index=b.index;
        const title=document.createElement('b');title.textContent=`P${b.index+1} · ${b.name||'Untitled beat'}`;
        const timing=document.createElement('span');timing.textContent=`${seconds(b.arrival)}${b.dwell_end>b.dwell_start?' · hold '+seconds(b.dwell_end-b.dwell_start):''}${b.cue?' · YOUR CUE':''}`;
        const w=(candidate&&el('dViewLabel').textContent==='Timing proposal'?candidate:original)?.waypoints?.[b.index];
        const angle=document.createElement('span');angle.textContent=w?`Pan ${w.yaw.toFixed(1)}° · Tilt ${w.pitch.toFixed(1)}°`:'';
        button.append(title,timing,angle);button.onclick=()=>{seek(b.arrival);selected=b.index;display();};el('dBeats').append(button);}
      const findings=shown.preflight.findings;el('dFindings').replaceChildren();
      el('dCheckTitle').textContent=shown.preflight.ok?'Sampled path check passed':'Resolve before shooting';
      for(const finding of findings){const item=document.createElement('li');
        const advice=finding.kind==='travel'?'Edit the framing or route; timing alone cannot fix travel.':finding.kind==='too fast'?'Allow more motion time or reduce angular travel.':finding.kind==='dead band'?'Try less time or more angular travel.':'Simplify the path for a bounded assessment.';
        item.textContent=`${finding.detail}. ${advice}`;el('dFindings').append(item);}
      el('dLimits').textContent=`Peaks: tilt ${shown.preflight.peak_pitch_dps.toFixed(2)}°/s · pan ${shown.preflight.peak_yaw_dps.toFixed(2)}°/s. ${shown.rig?.max_dps??data?.rig?.max_dps??'—'}°/s selected cap. ${shown.preflight.samples} assessment samples.`;
      for(const id of ['dPlay','dReset','dNext','dScrub','dEdit'])el(id).disabled=!shown.samples.length;
      seek(0);
    }
    async function analyse(target=null){
      const ticket=++requestId;clearAnalysis('Checking the authored path…');
      const draft=api.getMove();const result=await api.post('/api/director/preview',{move:draft,...target!==null?{target_duration:target}:{}});
      if(ticket!==requestId)return false;
      if(!result.ok){el('dState').textContent=result.error;return false;}
      data=shown=result.director;original=structuredClone(draft);candidate=data.proposal.move;
      el('dName').textContent=draft.name||'Untitled shot';el('dViewLabel').textContent='Authored path';
      el('dFit').disabled=el('dAuto').disabled=!data.samples.length;
      el('dState').textContent=data.samples.length?'Ready to preview. These controls never move the camera.':'Add at least two positions in Compose.';
      el('dProposal').hidden=data.proposal.status==='none';el('dChanges').replaceChildren();
      el('dProposalTitle').textContent=candidate?`${seconds(data.proposal.assessed_duration)} proposal`:'No checked timing available';
      el('dProposalText').textContent=candidate?(data.proposal.status==='suggested'?`Your requested timing did not pass. This alternative passed the sampled check. ${data.proposal.reason||''}`:'This timing passed the sampled check. Preview it before applying.'):(data.proposal.reason||'');
      if(candidate){for(let i=1;i<candidate.waypoints.length;i++){const li=document.createElement('li');li.textContent=`P${i} → P${i+1}: ${seconds(original.waypoints[i].duration)} → ${seconds(candidate.waypoints[i].duration)}`;el('dChanges').append(li);}}
      el('dApply').disabled=el('dPreviewProposal').disabled=!candidate;renderShown();return true;
    }
    el('dPlay').onclick=play;el('dReset').onclick=()=>seek(0);el('dScrub').oninput=()=>seek(+el('dScrub').value);
    el('dContinue').onclick=()=>{if(waiting!==null)consumed.add(waiting);waiting=null;el('dCue').hidden=true;play();};
    el('dNext').onclick=()=>{const next=shown?.beats.find(b=>b.arrival>t+.01);seek(next?next.arrival:end());};
    el('dRefresh').onclick=()=>analyse();el('dAuto').onclick=()=>analyse();
    el('dFit').onclick=()=>{const n=+el('dTarget').value;if(!el('dTarget').value||!Number.isFinite(n)||n<=0||n>86400){el('dState').textContent='Enter a duration between 0 and 86400 seconds.';el('dTarget').focus();return;}analyse(n);};
    el('dPreviewProposal').onclick=async()=>{if(!candidate)return;pause();const ticket=++requestId;
      const result=await api.post('/api/director/preview',{move:candidate});if(ticket!==requestId)return;
      if(!result.ok){el('dState').textContent=result.error;return;}shown=result.director;el('dViewLabel').textContent='Timing proposal';renderShown();play();};
    async function applyCandidate(proposed,selection=null){
      if(!proposed||applying)return false;pause();applying=true;const before=structuredClone(api.getMove());el('dApply').disabled=true;
      try{if(await api.apply(structuredClone(proposed),selection)){undo=before;if(await analyse()){el('dUndo').hidden=false;el('dState').textContent='Treatment applied to the draft. Camera unchanged.';}return true;}return false;}
      finally{applying=false;el('dApply').disabled=!candidate;}}
    el('dApply').onclick=()=>applyCandidate(candidate);
    el('dUndo').onclick=async()=>{if(!undo||applying)return;const previous=undo;pause();applying=true;
      try{if(await api.apply(previous)){undo=null;await analyse();el('dUndo').hidden=true;}}finally{applying=false;}};
    el('dEdit').onclick=()=>api.edit(selected);
    const resize=new ResizeObserver(()=>draw());resize.observe(el('dPlot'));
    document.addEventListener('visibilitychange',()=>{if(document.hidden)pause();});
    async function previewTreatment(treatment){
      const ticket=++requestId;clearAnalysis('Checking this AI treatment locally…');
      const result=await api.post('/api/director/preview',{move:treatment.move});
      if(ticket!==requestId||!result.ok)return false;
      data=shown=result.director;original=structuredClone(treatment.move);
      el('dName').textContent=treatment.title;el('dViewLabel').textContent='AI treatment · not applied';
      renderShown();el('dState').textContent='Rehearsing an AI timing treatment. Authored draft and camera unchanged.';
      el('dPlot').scrollIntoView({block:'center',behavior:'auto'});play();return true;
    }
    return {activate:()=>analyse(),pause,invalidate,previewTreatment,applyTreatment:applyCandidate};
  };
})();
